src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -188,12 +236,21 @@ CitusModifyBeginScan(CustomScanState *node, EState *estate, int eflags)
 
 	Job *workerJob = distributedPlan->workerJob;
 	Query *jobQuery = workerJob->jobQuery;
-	List *taskList = workerJob->taskList;
 
-	if (workerJob->requiresMasterEvaluation)
+	PlanState *planState = &(scanState->customScanState.ss.ps);
+	EState *executorState = planState->state;
+
+	/* citus only evaluates functions for modification queries */
+	bool modifyQueryRequiresMasterEvaluation =
+		workerJob->requiresMasterEvaluation && jobQuery->commandType != CMD_SELECT;
+
+	if (modifyQueryRequiresMasterEvaluation || workerJob->deferredPruning)
```

Comment:
deferredPruning can be removed from this if statement, since it's never true without modifyQueryRequiresMasterEvaluation being true.

Reply:
If that's an invariant the `workerJob` then that's not clear from the current code to me. Probably good to put an `Assert` at the top in that case.

Reply:
> deferredPruning can be removed from this if statement, since it's never true without modifyQueryRequiresMasterEvaluation being true.

With this PR, this could happen now. For fast-path router queries with parameters, we only defer the pruning (e.g., not `modifyQueryRequiresMasterEvaluation `).  And, `ExecuteMasterEvaluableFunctions` handles both function evaluation and parameter evaluation.

From performance wise, it seems good to traverse the query once, not splitting the function evaluation from parameter evaluation. 

> The branching in this function is quite complex I think adding the following would help:

The branching in the function doesn't look too complicated to me. I think the missing part was comments as it is not clear to the readers why we're evaluating the function when `deferredPruning`.  How does it look now @JelteF ?

Additionally, I can consider renaming `ExecuteMasterEvaluableFunctions` to something that explains we evaluate parameters as well.

Reply:
Comments definitely help indeed, but exiting early when possible makes the possible code paths much easier to understand. It also simplifies this check from effectively 3 boolean expressions to 2.

Reply:
I tried that, but didn't like much, I do want to get rid of anything related to commandType, it really shouldn't matter. The things that should matter are `workerJob->requiresMasterEvaluation` and `workerJob->deferredPruning`.

In the next release, I really want to get rid of `modifyQueryRequiresMasterEvaluation` just use `queryRequiresMasterEvaluation`

The current flow is not that complicated, we're calculating only two things:

1) Do we need to call `ExecuteMasterEvaluableFunctions`
2) Do we need to do pruning

I added an intermediate variable, and applied your below requested change